docs/rfcs/2025-07-07-node-deletion-api-improvement.md

```diff
@@ -0,0 +1,149 @@
+as part of the node stored information. Each node has a separate `NodeLifecycle` field with two
+possible states: `Active` and `Deleted`. When node deletion completes, the database row is not
+deleted but instead has its `NodeLifecycle` column switched to `Deleted`. Nodes with `Deleted`
+lifecycle are treated as if the row is absent for most handlers, with several exceptions: reattach
+and register functionality must be aware of tombstones. Additionally, new debug handlers are
+available for listing and deleting tombstones via the `/debug/v1/tombstone` path.
+
+### Gracefulness
+
+The problem of making node deletion graceful is complex and involves several challenges:
+
+- **Cancellable**: Since graceful deletion can be time-consuming, the operation must be cancellable
+to allow administrators to abort the process if needed, e.g. if run by mistake.
+- **Non-blocking**: Since graceful deletion can be time-consuming, we don't want to block
+deployment operations like draining/filling on the node deletion process. We need clear policies
+for handling concurrent operations: what happens when a drain/fill request arrives while deletion
+is in progress, and what happens when a delete request arrives while drain/fill is in progress.
+- **Persistent**: If the storage controller restarts during this long-running operation, we must
+preserve progress and automatically resume the deletion process after the storage controller
+restarts.
+- **Migrated correctly**: We cannot simply use the existing drain mechanism for nodes scheduled for
+deletion, as this would move shards to irrelevant locations. This could result in unnecessary load
+on the storage controller and inefficient resource utilization.
+- **Context-aware**: We need to improve our scheduling context awareness mechanism beyond the basic
+API to correctly calculate preferred node when using non-locked state operations.
+- **Force option**: Administrators need the ability to force immediate, non-graceful deletion when
+time constraints or emergency situations require it, bypassing the normal graceful migration
+process.
+
+The proposed solution addresses these challenges through the following components:
+
+- **New NodeLifecycle**: Introduce `NodeLifecycle::ScheduledForDeletion` to track nodes pending
+deletion. This state persists across storage controller restarts and prevents re-registration.
+- **New NodeSchedulingPolicy**: Introduce `NodeSchedulingPolicy::Deleting` to track the current
+deletion process if any.
+- **OperationTracker**: A service state object that manages all long-running operations with the
+following capabilities:
+  - Execute operations (drain/delete/fill) with proper concurrency control
+  - Cancel ongoing operation
+  - Queue deletion intents when there are another running operations
+  - Enforce operation rules (e.g., only one drain/fill operation at a time)
+  - Maintain persistent intent tracking for queued deletions
```

Comment:
"persistent intent" is confusing to me. Intent state is by definition not persisted.
I would side-step all of this deletion queuing stuff by not allowing the deletion to start while there's another ongoing operation (also don't allow it to start in between: `NodeSchedulingPolicy::PauseForRestart`. If one really need to delete the node, they can use a force flag.